### DISCUSSION TITLE
Operator on tuple

### DISCUSSION THREAD

**Discussion by 20chan** [dotnet/csharplang#8751] [Language Ideas]
Like Nullable<T> supports operators defined at T, Tuple can be. 

Example:

```csharp
void DoSomething((int x, int y) c1, (int x, int y) c2)
{
    int (a, b) = (c1.x + c2.x, c1.y + c2.y);
    (a, b) = (-a, -b);
}
```

Could be

```csharp
void DoSomething((int x, int y) c1, (int x, int y) c2)
{
    int (a, b) = c1 + c2;
    (a, b) = -(a, b);
}
```

Even if types of all elements in tuple is not identical operator can be used if all types override specific operator.

```csharp
(int a, string b) = (1, "Hello, ") + (10, "World!");
```

**Comment by bondsbw** [comment]
Would applying the operator to tuples of different lengths be allowed?  If so, what would be the runtime behavior?

**Comment by 20chan** [comment]
Forgot that one. Lengths of tuples must be same and it will be checked when compiling.

**Comment by bondsbw** [comment]
Related: #190

**Comment by stakx** [comment]
Which binary operators are you thinking of here? Just `+` and `-`, or also others like `&`, `|`, `^`, `%`?

If this makes it into the language, it might not take long until someone suggests `k * (a, b, c , ...)` where `k` is a single value. Should this be allowed too and translated as `(k * a, k * b, k * c, ...)`?

**Comment by Unknown6656** [comment]
@stakx : Yes, please!!
And IMO it should be allowed on all binary and unary operators.

**Comment by quinmars** [comment]
@stakx and what happens if the single value of `k` is `(1, "Hello, ")`?  😄

**Comment by stakx** [comment]
@quinmars - I think you know how I meant it. 😆 But for the record, let's say "where `k` is not a tuple".

In fact, with the above I wanted to show that this proposal for operator `+` on tuples could quickly grow into something bigger, and not every other operator's workings might be as easy to decide as `+`. For example, would there be any real-world use for short-circuiting operators such as `&&` applied to tuples? What would unambiguous syntax look like for the application of some `explicit operator T(<tuple type>)`? etc.

I do like this proposal, but I can't help but wonder whether the request for these kind of operations on tuples have a mostly numerical / mathematical background, in which case it might be better to use `System.Numerics.Vectors` et al. instead of tuples...?

**Comment by bondsbw** [comment]
A tuple is not a mathematical vector, and we should be careful not to confuse the two.

**Comment by jnm2** [comment]
This is in line with the principle that operations on a tuple should be distributive to its elements. I like it.
